src/include/type/transient_value.h

```diff
@@ -0,0 +1,282 @@
+ * class for values between the parser layer and the execution layer. TransientValue objects are created using the
+ * TransientValueFactory, rather than exposing any constructors directly. This is to make it as obvious as possible if
+ * the user is doing something that may hurt performance like repeatedly creating and destroying TransientValue objects.
+ * C types are extracted from TransientValue objects using the TransientValuePeeker class.
+ * @see TransientValueFactory
+ * @see TransientValuePeeker
+ */
+class TransientValue {
+  friend class TransientValueFactory;                     // Access to constructor
+  friend class TransientValuePeeker;                      // Access to GetAs
+  friend class terrier::parser::ConstantValueExpression;  // Access to copy constructor
+
+ public:
+  /**
+   * @return TypeId of this TransientValue object.
+   */
+  TypeId Type() const {
+    // bitwise AND the TypeId with 01111111 to return TypeId value without the embedded NULL bit
+    return static_cast<TypeId>(static_cast<uint8_t>(type_) & 0x7F);
+  }
+
+  /**
+   * Can't see any reason to have a default constructor. If we were to have one, it would probably just set the TypeId
+   * to boolean and SetNull(true) but I can't come up with a good reason to expose this.
+   */
+  TransientValue() = delete;
+
+  /**
+   * @return true if TransientValue is a SQL NULL, otherwise false
+   */
+  bool Null() const {
+    // bitwise AND the TypeId with 1000000 to extract NULL bit
+    return static_cast<bool>(static_cast<uint8_t>(type_) & 0x80);
+  }
+
+  /**
+   * Change the SQL NULL value of this TransientValue. We use the MSB to reflect this since we don't need all 8 bits for
+   * TypeId
+   * @param set_null true if TransientValue should be set to NULL, false otherwise
+   */
+  void SetNull(const bool set_null) {
+    if (set_null) {
+      // bitwise OR the TypeId with 1000000 to set NULL bit
+      type_ = static_cast<TypeId>(static_cast<uint8_t>(type_) | 0x80);
+    } else {
+      // bitwise AND the TypeId with 01111111 to clear NULL bit
+      type_ = static_cast<TypeId>(static_cast<uint8_t>(type_) & 0x7F);
+    }
+  }
+
+  /**
+   * VARCHAR's own their own internal buffer so we need to make sure to free it.
+   */
+  ~TransientValue() {
+    if (Type() == TypeId::VARCHAR) {
+      delete[] reinterpret_cast<char *const>(data_);
+    }
+  }
+
+  /**
+   * @param rhs TransientValue to compare this against
+   * @return true if two TransientValues are equal, satisfying the following predicates:
+   * 1) they share the same TypeId
+   * 2) they share the same NULL value
+   * 3) they share the same data_ value if non-VARCHARs, or their VARCHAR contents are the same if VARCHARS
+   * false otherwise
+   */
+  bool operator==(const TransientValue &rhs) const {
+    if (type_ != rhs.type_) return false;  // checks TypeId and NULL at the same time due to stolen MSB of type_ field
```

Comment:
I believe in SQL, null == null should either be false or null.

Reply:
These are following object equality semantics and this operator wouldn't be used for SQL comparisons (like if the parser needs to check equality of nodes with contained TransientValues). If you think that's incorrect we can talk about it, but the idea would be SQL predicate evaluations are only done in the index comparators which will be handled elsewhere.

Reply:
That's reasonable. Let's note that explicitly in the doc though that this is not equivalent to the SQL equal operator

Reply:
Should we even include `==` then? I am thinking about collations. 

When would anybody actually need to compare to `TransientValues` to see whether they are equivalent?